Fix RPCTimerInterface ordering issue

jonasschnelli opened the pull request:
Dispatching a `QThread` from a non Qt thread is not allowed. Always use the `HTTPRPCTimerInterface` (non QT) to dispatch `RPCRunLater` threads.

Agree that using the order of `timerInterfaces` is not the best way of identifying the type of the timer interface.

fixes #7316

jonasschnelli committed:
Fix RPCTimerInterface ordering issue
Dispatching a QThread from a non Qt thread is not allowed. Always use the HTTPRPCTimerInterface (non QT) to dispatch RPCRunLater threads.

MarcoFalke commented:
Tested ACK 9490ec3, indeed fixes https://github.com/bitcoin/bitcoin/issues/7316#issuecomment-169938133.

Tag "Backport" missing?

jonasschnelli commented:
Discussed with @laanwj on IRC and i'll try now to remove the set of timers, I think having only one timer makes things more stable for future changes.

laanwj commented:
Concept ACK, but the change in logic needs to be more explicit. This is too fragile against refactors/reorders (at least for master, for 0.12 as a quick last minute patch it'd be acceptable...).

jonasschnelli committed:
[RPC] remove the option of having multiple timer interfaces

jonasschnelli commented:
Added a commit that removes the `std::vector` for RPCTimerInterfaces (just one pointer now). Also added a function (`RPCSetTimerInterfaceIfUnset`) that prevents setting a Qt bases timer interface if already only is set.
Not meant for back porting (will open "the" simple fix for 0.12).

laanwj commented on src/rpcserver.h:
```diff
@@ -100,10 +100,12 @@ class RPCTimerInterface
-void RPCUnregisterTimerInterface(RPCTimerInterface *iface);
+/** Set the factory function for timers */
+void RPCSetTimerInterface(RPCTimerInterface *iface);
+/** Set the factory function for timer, but only, it unset */
```
typo: if unset

jonasschnelli replied:
Thanks. Fixed.

laanwj commented:
Nice, uTACK

laanwj commented on src/qt/rpcconsole.cpp:
```diff
@@ -278,7 +278,9 @@ RPCConsole::RPCConsole(const PlatformStyle *platformStyle, QWidget *parent) :
     // Register RPC timer interface
     rpcTimerInterface = new QtRPCTimerInterface();
-    RPCRegisterTimerInterface(rpcTimerInterface);
+    // avoid accidentailly overwrite an existing, non QTThread
```
s/accidentiailly overwrite/accidentally overwriting

jonasschnelli replied:
Thanks. Force push fixed.
